### PR TITLE
Add stricter ESLint rules for Typescript files

### PR DESCRIPTION
**Pull request by renchap**
My goal with this PR is to provide more strict linting of TS files to have a more consistent code-style, automatically enforced.

## Main config changes

- ESLint on TS files is now [type-aware](https://typescript-eslint.io/linting/typed-linting), allowing more rules to be enabled
- ESLint Import plugin was using the `node` resolver, but we should be using the Typescript one, so our aliases defined in `tsconfig.json` are understood by eslint

## Small rules changes & explanation

- `import/consistent-type-specifier-style` to prefer defining types with `interface` when possible, this makes it consistent as otherwise you would have an `interface` when you `extend`, and a `type` otherwise` (common with React props)
- `@typescript-eslint/consistent-type-{imports,exports}`: this is recommended as it helps with bundling, tree-shaking and possible circular dependencies issues
- `import/consistent-type-specifier-style` to always have `import type { T }` and no `import { type T }`
- enabled back `@typescript-eslint/no-explicit-any` and fixed it, this should be disabled only when no other choice (migration)
- `import/no-anonymous-default-export` to force default exports to have a name
- `import/no-self-import` as this can sometimes happen with refactors
- `import/no-useless-path-segments` to ensure we have the smallest relative path possible

## Enforce the `import` statements order

This makes the files more readable, allowing for an easy scanning of what is imported and makes it faster to find a specific import.

They will be grouped by types, then sorted alphabeticaly in each group.

The groups will be:
- Internal imports (NodeJS)
- React / ReactDOM / PropTypes
- React Intl
- Immutable / Redux
- absolute `mastodon/` imports
- relative imports

I chose to group the common imports in their own groups, so they can be more easily scanned.

This is auto-fixed by ESLint, so it does not bring any developer overhead.

I chose to only apply it to Typescript files as there are only a few of them for now, but this could be applied to all JS files if wanted.

If applied to every files, this would result in this change: `364 files changed, 2462 insertions(+), 1324 deletions(-)`.
This could be applied later, or I can add this commit to the PR, but I dont want to cause a huge rebase work for others PRs or forks (but thats an autofix, so its ok?), please tell me if I should do it or not.

### CODE REVIEW

**Review by ykzts — APPROVED** [review]

**Comment by renchap** [comment]
After discussing with @ClearlyClaire I will send a PR later with a proposal for rearranging the `import` statements in the whole codebase, so its fine merging this one with only the TS files for now, if you are ok with it.

**Comment by renchap** [comment]
I added a few comments in the eslint config for `import/order`, and a few other packages to be grouped.

FYI, here are the external packages imported more than 1 time in the codebase (I used this to see what should be grouped):

```
   2 fuzzysort
   2 history
   2 react-router-scroll-4
   2 react-sparklines
   2 stacktrace-js
   3 intl-messageformat
   3 react-swipeable-views
   3 react-textarea-autosize
   3 stringz
   4 axios
   4 escape-html
   4 react-redux-loading-bar
   4 react-toggle
   5 react-hotkeys
   5 react-test-renderer
   7 react-dom
   8 detect-passive-events
  11 reselect
  29 lodash
  32 react-helmet
  32 react-router-dom
  62 classnames
  72 immutable
  75 react-immutable-pure-component
 108 react-immutable-proptypes
 140 react-redux
 179 react-intl
 194 prop-types
 243 react
```

File: .eslintrc.js [hunk 1]
```diff
@@ -257,7 +260,9 @@ module.exports = {
 
       extends: [
         'eslint:recommended',
+        'plugin:@typescript-eslint/eslint-recommended',
```
Comment (nschonni) [inline .eslintrc.js:263]: What's the difference in this one from the ones right before and right after?
It looks like it's already contained in the later line as well https://github.com/typescript-eslint/typescript-eslint/blob/a6f23a2dd5f1ad8e3ede0a4594a5e435182b4a2c/packages/eslint-plugin/src/configs/recommended-requiring-type-checking.ts#L6

Comment (renchap) [reply .eslintrc.js:263]: Good point, this is probably a mistake due to my various rebases. `plugin:@typescript-eslint/eslint-recommended` is not needed here, the other 2 are the recommended configuration [from their doc](https://typescript-eslint.io/linting/configs#recommended-requiring-type-checking)

Comment (nschonni) [reply .eslintrc.js]: Interesting, looks like there is a "hard mode" too https://typescript-eslint.io/linting/configs/#strict

**Pull request merged**